### PR TITLE
Remotetap extension concept

### PR DESCRIPTION
**Pull request by wildum**
#### Description

This is the core part of the remote tap concept described here: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/34096

This will be used by the new version of the [remotetap extension in contrib](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/34872).

This new version uses the processorhelper to tap into the processor pipeline. ALL processors using the processorhelper to be created will automatically support this feature.

To see what a full version could look like you can check this POC: https://github.com/open-telemetry/opentelemetry-collector/pull/10962

When no remotetap stream is open, the additional work on the processor pipeline is negligible (1 map read / proc consumption if the extension is not enabled, 2 map read / proc consumption if the extension is enabled but no streams are open)

Idealy, in the future, the feature would be available to all kinds of components. Taping into the processors that use the processorhelper pkg is the first easy step in this direction.

### CODE REVIEW

**Comment by atoulme** [comment]
Please create this PR in contrib? We will not have new components in this repository right now.

**Comment by wildum** [comment]
Ah okay, then I will move the component back to contrib and will only update the processorhelper logic in core

**Comment by wildum** [comment]
@atoulme I moved the component from this PR to contrib: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/34872.
Only the changes related to the framework changes are left in this PR

File: service/internal/graph/graph.go [hunk 1]
```diff
@@ -49,6 +49,8 @@ type Settings struct {
 	PipelineConfigs pipelines.Config
 
 	ReportStatus status.ServiceStatusFunc
+
+	Extensions func() map[component.ID]component.Component
```
[the remark this answers is not in the document]
Comment (wildum) [reply service/internal/graph/graph.go:53]: Thanks, that's great feedback, I will try to come up with something next week :)

**Review by djaglowski — COMMENTED** [review]
This looks pretty good to me. I suspect the most important thing to move this forward is to make sure it does not in any way interfere with the effort to mark modules 1.0, since this feature would necessarily be experimental for some time. If I am not mistaken, this means the changes to `pdata` are likely problematic. Can we move them elsewhere, maybe to `processorhelper`?

**Comment by wildum** [comment]
> Can we move them elsewhere, maybe to processorhelper?

I could not move it to processorhelper because it causes some cyclic dep problems + ideally in the future I would like this interface by other types of components. I moved it to the component pkg, what do you think? IMO it fits great there, next to the telemetry.go that provides components the ability to report telemetry

**Comment by djaglowski** [comment]
Unfortunately I think we have the same stability concern with `component`. Basically, anything mentioned [here](https://github.com/open-telemetry/opentelemetry-collector/issues/9375) as part of the roadmap is not going to work for something so new. Maybe @open-telemetry/collector-approvers have an idea.

**Comment by wildum** [comment]
Moved the publisher interface to the processor/internal pkg. It's not ideal because in the future it would be nice to expand the functionality to other components but that should be ok for a first step

**Comment by wildum** [comment]
@djaglowski I like your approach :) Helpers should be used to provide optional features for components that require special behaviors (for example "WithStart", "WithCapabilities" and "WithShutdown"). Forcing all components to use the helpers by putting features such as the telemetry or the remote tap in the helpers defeats the purpose of the helpers. Such functionalities should be rooted at a deeper level and make use of the graph structure.

I'm happy to close my PR and support yours. The bad news is that it's not great timing with the 1.0 work as you mentioned above but the good news is that afaik the modifications don't introduce breaking changes (as long as we keep the helpers)?

**Comment by djaglowski** [comment]
> The bad news is that it's not great timing with the 1.0 work

I believe the changes are isolated to the service package, which is not part of the 1.0 effort, if I understand correctly. :)

**Pull request closed without merging**